filterExprParser.go

```diff
@@ -0,0 +1,691 @@
+		} else {
+			output = append(output, "OR")
+		}
+		output = append(output, expr.String())
+	}
+
+	return strings.Join(output, " ")
+}
+
+type FEParen struct {
+	OpenParen  *FEOpenParen  `@@ |`
+	CloseParen *FECloseParen `@@`
+}
+
+func (fep *FEParen) IsOpen() bool {
+	return fep.OpenParen != nil && fep.CloseParen == nil
+}
+
+func (fep *FEParen) IsClose() bool {
+	return fep.CloseParen != nil && fep.OpenParen == nil
+}
+
+func (fep *FEParen) String() string {
+	if fep.IsOpen() {
+		return fep.OpenParen.String()
+	} else if fep.IsClose() {
+		return fep.CloseParen.String()
+	} else {
+		return "?? (FEParen)"
+	}
+}
+
+type FEOpenParen struct {
+	Parens string `@"("`
+}
+
+func (feop *FEOpenParen) String() string {
+	return "("
+}
+
+type FECloseParen struct {
+	Parens string `@")"`
+}
+
+func (fecp *FECloseParen) String() string {
+	return ")"
+}
+
+type FEAndCondition struct {
+	OrConditions []*FECondition `@@ { "AND" @@ }`
+}
+
+func (ac *FEAndCondition) String() string {
+	output := []string{}
+
+	first := true
+	for _, e := range ac.OrConditions {
+		if first {
+			first = false
+		} else {
+			output = append(output, "AND")
+		}
+		output = append(output, e.String())
+	}
+
+	return strings.Join(output, " ")
+}
+
+type FECondition struct {
+	PreParen  []*FEParen   `{ @@ }`
+	Not       *FECondition `"NOT" @@`
+	Operand   *FEOperand   `| @@`
+	PostParen []*FEParen   `{ @@ }`
+}
+
+func (fec *FECondition) String() string {
+	var outputStr []string
+
+	for i := 0; i < len(fec.PreParen); i++ {
+		outputStr = append(outputStr, fec.PreParen[i].String())
+	}
+
+	if fec.Not != nil {
+		outputStr = append(outputStr, fmt.Sprintf("NOT %v", fec.Not.String()))
+	} else if fec.Operand != nil {
+		outputStr = append(outputStr, fec.Operand.String())
+	} else {
+		outputStr = append(outputStr, "?? (FECondition)")
+	}
+
+	for i := 0; i < len(fec.PostParen); i++ {
+		outputStr = append(outputStr, fec.PostParen[i].String())
+	}
+
+	return strings.Join(outputStr, " ")
+}
+
+type FEOperand struct {
+	BooleanExpr *FEBooleanExpr `@@ |`
+	LHS         *FELhs         `( @@ (`
+	Op          *FECompareOp   `( @@`
+	RHS         *FERhs         `@@ ) | `
+	CheckOp     *FECheckOp     `@@ ) )`
+}
+
+func (feo *FEOperand) String() string {
+	if feo.BooleanExpr != nil {
+		return feo.BooleanExpr.String()
+	} else if feo.CheckOp != nil {
+		return fmt.Sprintf("%v %v", feo.LHS.String(), feo.CheckOp.String())
+	} else if feo.Op != nil {
+		return fmt.Sprintf("%v %v %v", feo.LHS.String(), feo.Op.String(), feo.RHS.String())
+	} else {
+		return "?? (FEOperand)"
+	}
+}
+
+type FEBooleanExpr struct {
+	BooleanVal  *FEBoolean         `@@ |`
+	BooleanFunc *FEBooleanFuncExpr `@@`
+}
+
+func (be *FEBooleanExpr) String() string {
+	if be.BooleanVal != nil {
+		return be.BooleanVal.String()
+	} else if be.BooleanFunc != nil {
+		return be.BooleanFunc.String()
+	} else {
+		return "?? (FEBooleanExpr)"
+	}
+}
+
+type FEBoolean struct {
+	TVal  *bool `@"TRUE" |`
+	TVal1 *bool `@"true" |`
```

Comment:
is the intention to keep user's original input so that we can print it later? what about "True/False"?

if printing is not user visible / critical, would it be simpler to keep only one "true/false" value in the struct?

Reply:
It's not for printing, but rather capturing/matching. We're only allowing the matching of either TRUE or true.